src/engine/GroupBy.cpp

```diff
@@ -759,57 +760,162 @@ GroupBy::checkIfHashMapOptimizationPossible(
   }
 
   // Make sure that there are no nested aggregates
-  for (auto& aliasAggregate : aliasAggregates) {
-    for (auto& info : aliasAggregate.aggregateInfo_) {
-      auto aggregateExpression = info.expr_;
+  for (auto& alias : aliasesWithAggregateInfo) {
+    for (auto& aggregateInfo : alias.aggregateInfo_) {
+      auto aggregateExpression = aggregateInfo.expr_;
 
       if (aggregateExpression->children().front()->containsAggregate())
         return std::nullopt;
     }
   }
 
-  // Make sure there are aggregates
-  if (numAggregates == 0) return std::nullopt;
-
   const Variable& groupByVariable = _groupByVariables.front();
   auto child = _subtree->getRootOperation()->getChildren().at(0);
   auto columnIndex = child->getVariableColumn(groupByVariable);
 
-  return HashMapOptimizationData{columnIndex, aliasAggregates, numAggregates};
+  return HashMapOptimizationData{columnIndex, aliasesWithAggregateInfo,
+                                 numAggregates};
+}
+
+// _____________________________________________________________________________
+std::optional<std::vector<GroupBy::HashMapAggregateInformation>>
+GroupBy::findAggregates(sparqlExpression::SparqlExpression* expr) {
+  std::vector<HashMapAggregateInformation> result;
+  if (!findAggregatesImpl(nullptr, expr, std::nullopt, result))
+    return std::nullopt;
+  else
+    return result;
+}
+
+// _____________________________________________________________________________
+template <class T>
+bool GroupBy::hasType(sparqlExpression::SparqlExpression* expr) {
+  return dynamic_cast<T*>(expr) != nullptr;
+}
+
+// _____________________________________________________________________________
+bool GroupBy::isSupportedAggregate(sparqlExpression::SparqlExpression* expr) {
+  using namespace sparqlExpression;
+
+  return !(hasType<SumExpression>(expr) || hasType<MinExpression>(expr) ||
+           hasType<MaxExpression>(expr) || hasType<CountExpression>(expr) ||
+           hasType<GroupConcatExpression>(expr));
 }
 
 // _____________________________________________________________________________
-bool GroupBy::findAggregateMultiple(
+bool GroupBy::findAggregatesImpl(
     sparqlExpression::SparqlExpression* parent,
     sparqlExpression::SparqlExpression* expr, std::optional<size_t> index,
     std::vector<GroupBy::HashMapAggregateInformation>& info) {
   if (dynamic_cast<sparqlExpression::AvgExpression*>(expr) != nullptr) {
-    info.push_back({expr, parent, index});
+    if (parent != nullptr && index.has_value()) {
+      ParentAndChildIndex parentAndChildIndex{parent, index.value()};
+      info.emplace_back(expr, 0, parentAndChildIndex);
+    } else
+      info.emplace_back(expr, 0, std::nullopt);
```

Comment:
Just for clarity, where is the check, that an aggregate itself may not have other aggregates as children? I think this is a natural position to check it here, or at least comment on where this will be checked.